File: docker/openclaw-runner/adapter/run_task.mjs
#!/usr/bin/env node
import { createWriteStream } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";
import { spawn } from "node:child_process";

const ADAPTER_NAME = "nanoclaw-openclaw-adapter";
const CAPTURE_LIMIT_BYTES = 2_000_000;

async function main() {
  const options = parseArgs(process.argv.slice(2));
  await ensureDir(options.output);
  await ensureDir(options.state);
  await ensureDir(options.workspace);

  const tracePath = path.join(options.output, "trace.jsonl");
  const startedAt = Date.now();
  const taskText = await fs.readFile(options.task, "utf8");
  const inputDir = path.dirname(options.task);
  const runnerRequest = await readJsonIfExists(
    path.join(inputDir, "runner_request.json"),
    {},
  );
  const priorMessages = await readJsonIfExists(
    path.join(inputDir, "prior_messages.json"),
    [],
  );
  const resolvedTask = await readJsonIfExists(options.resolvedTask, {});

  const env = await buildOpenClawEnv(options, runnerRequest);
  const message = buildMessage(taskText, priorMessages, env);
  const sessionId = resolveSessionId(runnerRequest, resolvedTask, env);
  const command = env.NANOCLAW_OPENCLAW_CLI || "openclaw";
  const args = buildOpenClawArgs(message, sessionId, env);

  await appendTrace(tracePath, {
    type: "adapter_started",
    adapter: ADAPTER_NAME,
    task_id: runnerRequest.task_id ?? resolvedTask.id ?? null,
    run_id: runnerRequest.run_id ?? null,
    turn: runnerRequest.turn ?? null,
    session_id: sessionId,
  });

  const onboarding = await maybeOnboardOpenClaw({
    command,
    env,
    workspace: options.workspace,
    outputDir: options.output,
    tracePath,
  });

  const stdoutPath = path.join(options.output, "openclaw_stdout.log");
  const stderrPath = path.join(options.output, "openclaw_stderr.log");

  await appendTrace(tracePath, {
    type: "openclaw_started",
    command: redactCommandForTrace(command, args),
    cwd: options.workspace,
  });

  const result = await runCommand(command, args, {
    cwd: options.workspace,
    env,
    stdoutPath,
    stderrPath,
  });

  const extraction = extractFinalAnswer(result.stdout, result.stderr);
  const finalAnswer = ensureTrailingNewline(extraction.content);
  await fs.writeFile(path.join(options.output, "final_answer.md"), finalAnswer, "utf8");

  const metadata = {
    adapter: ADAPTER_NAME,
    framework: "openclaw",
    command: redactCommandForTrace(command, args),
    cwd: options.workspace,
    session_id: sessionId,
    exit_code: result.exitCode,
    signal: result.signal,
    error: result.error,
    final_answer_strategy: extraction.strategy,
    final_answer_bytes: Buffer.byteLength(finalAnswer, "utf8"),
    stdout_file: path.basename(stdoutPath),
    stderr_file: path.basename(stderrPath),
    onboarding,
    duration_ms: Date.now() - startedAt,
  };
  await fs.writeFile(
    path.join(options.output, "runner_metadata.json"),
    JSON.stringify(metadata, null, 2) + "\n",
    "utf8",
  );

  await appendTrace(tracePath, {
    type: "openclaw_finished",
    exit_code: result.exitCode,
    signal: result.signal,
    error: result.error,
    final_answer_strategy: extraction.strategy,
    duration_ms: Date.now() - startedAt,
  });

  process.exitCode = normalizeExitCode(result.exitCode, result.signal);
}

function parseArgs(argv) {
  const options = {};
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--help" || arg === "-h") {
      printUsageAndExit();
    }
    if (!arg.startsWith("--")) {
      throw new Error(`Unexpected argument: ${arg}`);
    }
    const key = arg.slice(2);
    const value = argv[index + 1];
    if (value === undefined || value.startsWith("--")) {
      throw new Error(`Missing value for ${arg}`);
    }
    index += 1;
    if (key === "resolved-task") {
      options.resolvedTask = value;
    } else if (["task", "workspace", "output", "state"].includes(key)) {
      options[key] = value;
    } else {
      throw new Error(`Unknown argument: ${arg}`);
    }
  }

  for (const key of ["task", "resolvedTask", "workspace", "output", "state"]) {
    if (!options[key]) {
      throw new Error(`Missing required argument: --${toKebabCase(key)}`);
    }
  }

  return {
    task: path.resolve(options.task),
    resolvedTask: path.resolve(options.resolvedTask),
    workspace: path.resolve(options.workspace),
    output: path.resolve(options.output),
    state: path.resolve(options.state),
  };
}

function printUsageAndExit() {
  process.stdout.write(
    [
      "Usage: run_task --task /input/task.md --resolved-task /input/resolved_task.json \\",
      "  --workspace /workspace --output /output --state /state",
      "",
    ].join("\n"),
  );
  process.exit(0);
}

function toKebabCase(value) {
  return value.replace(/[A-Z]/g, (match) => `-${match.toLowerCase()}`);
}

async function buildOpenClawEnv(options, runnerRequest) {
  const env = { ...process.env };
  if (!env.NANOCLAW_MODEL && runnerRequest.model) {
    env.NANOCLAW_MODEL = String(runnerRequest.model);
  }
  if (env.NANOCLAW_BASE_URL && !env.OPENAI_BASE_URL) {
    env.OPENAI_BASE_URL = env.NANOCLAW_BASE_URL;
  }
  if (env.NANOCLAW_BASE_URL && !env.CUSTOM_API_KEY && env.OPENAI_API_KEY) {
    env.CUSTOM_API_KEY = env.OPENAI_API_KEY;
  }

  if (env.NANOCLAW_OPENCLAW_ISOLATED_HOME !== "0") {
    const homeDir = env.NANOCLAW_OPENCLAW_HOME || path.join(options.state, "home");
    const xdgDir = path.join(options.state, "xdg");
    env.HOME = homeDir;
    env.XDG_CONFIG_HOME = env.XDG_CONFIG_HOME || path.join(xdgDir, "config");
    env.XDG_STATE_HOME = env.XDG_STATE_HOME || path.join(xdgDir, "state");
    env.XDG_CACHE_HOME = env.XDG_CACHE_HOME || path.join(xdgDir, "cache");
    await ensureDir(homeDir);
    await ensureDir(env.XDG_CONFIG_HOME);
    await ensureDir(env.XDG_STATE_HOME);
    await ensureDir(env.XDG_CACHE_HOME);
  }

  return env;
}

function buildMessage(taskText, priorMessages, env) {
  if (
    env.NANOCLAW_OPENCLAW_INCLUDE_PRIOR_MESSAGES !== "1"
    || !Array.isArray(priorMessages)
    || priorMessages.length === 0
  ) {
    return taskText;
  }

  const history = priorMessages
    .filter((message) => message && typeof message === "object")
    .map((message) => {
      const role = String(message.role ?? "message").toUpperCase();
      const content = String(message.content ?? "");
      return `[${role}]\n${content}`;
    })
    .filter((message) => message.trim().length > 0)
    .join("\n\n");

  if (!history) {
    return taskText;
  }

  return `${history}\n\n[USER]\n${taskText}`;
}

function resolveSessionId(runnerRequest, resolvedTask, env) {
  const raw =
    env.NANOCLAW_OPENCLAW_SESSION_ID
    || [
      "nanoclaw",
      runnerRequest.task_id ?? resolvedTask.id ?? "task",
      runnerRequest.run_id ?? "run",
    ].join("-");
  return String(raw).replace(/[^a-zA-Z0-9_.:-]+/g, "-").slice(0, 180);
}

function buildOpenClawArgs(message, sessionId, env) {
  const args = ["agent", "--local", "--message", message, "--json"];

  if (env.NANOCLAW_OPENCLAW_AGENT) {
    args.push("--agent", env.NANOCLAW_OPENCLAW_AGENT);
  } else {
    args.push("--session-id", sessionId);
  }

  if (env.NANOCLAW_MODEL) {
    args.push("--model", env.NANOCLAW_MODEL);
  }
  if (env.NANOCLAW_OPENCLAW_THINKING) {
    args.push("--thinking", env.NANOCLAW_OPENCLAW_THINKING);
  }
  if (env.NANOCLAW_OPENCLAW_VERBOSE) {
    const verboseValue = env.NANOCLAW_OPENCLAW_VERBOSE === "1"
      ? "on"
      : env.NANOCLAW_OPENCLAW_VERBOSE === "0"
        ? "off"
        : env.NANOCLAW_OPENCLAW_VERBOSE;
    args.push("--verbose", verboseValue);
  }
  if (env.NANOCLAW_OPENCLAW_TIMEOUT) {
    args.push("--timeout", env.NANOCLAW_OPENCLAW_TIMEOUT);
  }

  return args;
}

async function maybeOnboardOpenClaw({ command, env, workspace, outputDir, tracePath }) {
  if (env.NANOCLAW_OPENCLAW_AUTO_ONBOARD === "0") {
    await appendTrace(tracePath, { type: "openclaw_onboard_skipped", reason: "disabled" });
    return { attempted: false, reason: "disabled" };
  }

  const configPath = path.join(env.HOME || "", ".openclaw", "openclaw.json");
  if (!env.HOME || await fileExists(configPath)) {
    const reason = env.HOME ? "config_exists" : "home_unset";
    await appendTrace(tracePath, { type: "openclaw_onboard_skipped", reason });
    return { attempted: false, reason };
  }

  const onboardArgs = buildOnboardArgs(env, workspace);
  if (!onboardArgs) {
    await appendTrace(tracePath, {
      type: "openclaw_onboard_skipped",
      reason: "no_supported_provider_env",
    });
    return { attempted: false, reason: "no_supported_provider_env" };
  }

  const stdoutPath = path.join(outputDir, "openclaw_onboard_stdout.log");
  const stderrPath = path.join(outputDir, "openclaw_onboard_stderr.log");

  await appendTrace(tracePath, {
    type: "openclaw_onboard_started",
    command: redactCommandForTrace(command, onboardArgs),
  });

  const result = await runCommand(command, onboardArgs, {
    cwd: env.HOME,
    env,
    stdoutPath,
    stderrPath,
  });

  await appendTrace(tracePath, {
    type: "openclaw_onboard_finished",
    exit_code: result.exitCode,
    signal: result.signal,
    error: result.error,
  });

  return {
    attempted: true,
    exit_code: result.exitCode,
    signal: result.signal,
    error: result.error,
    stdout_file: path.basename(stdoutPath),
    stderr_file: path.basename(stderrPath),
  };
}

function buildOnboardArgs(env, workspace) {
  const baseArgs = [
    "onboard",
    "--non-interactive",
    "--mode",
    "local",
    "--workspace",
    workspace,
    "--secret-input-mode",
    "ref",
    "--accept-risk",
    "--no-install-daemon",
    "--skip-daemon",
    "--skip-channels",
    "--skip-health",
    "--skip-search",
    "--skip-skills",
    "--skip-ui",
  ];

  if (env.NANOCLAW_BASE_URL && env.NANOCLAW_MODEL && env.CUSTOM_API_KEY) {
    return [
      ...baseArgs,
      "--auth-choice",
      "custom-api-key",
      "--custom-base-url",
      env.NANOCLAW_BASE_URL,
      "--custom-model-id",
      env.NANOCLAW_MODEL,
      "--custom-compatibility",
      env.NANOCLAW_OPENCLAW_CUSTOM_COMPATIBILITY || "openai",
    ];
  }

  if (env.OPENAI_API_KEY) {
    return [...baseArgs, "--auth-choice", "openai-api-key"];
  }

  return null;
}

async function runCommand(command, args, { cwd, env, stdoutPath, stderrPath }) {
  await ensureDir(path.dirname(stdoutPath));
  await ensureDir(path.dirname(stderrPath));

  return new Promise((resolve) => {
    const stdout = makeStreamCollector(stdoutPath);
    const stderr = makeStreamCollector(stderrPath);
    let spawnError = null;

    let child;
    try {
      child = spawn(command, args, {
        cwd,
        env,
        stdio: ["ignore", "pipe", "pipe"],
      });
    } catch (error) {
      spawnError = error;
      Promise.all([stdout.close(), stderr.close()]).then(() => {
        resolve({
          exitCode: 127,
          signal: null,
          error: error.message,
          stdout: "",
          stderr: error.message,
        });
      });
      return;
    }

    child.stdout.on("data", (chunk) => stdout.write(chunk));
    child.stderr.on("data", (chunk) => stderr.write(chunk));
    child.on("error", (error) => {
      spawnError = error;
    });
    child.on("close", (exitCode, signal) => {
      Promise.all([stdout.close(), stderr.close()]).then(() => {
        resolve({
          exitCode: spawnError ? 127 : exitCode,
          signal,
          error: spawnError ? spawnError.message : null,
          stdout: stdout.text,
          stderr: stderr.text,
        });
      });
    });
  });
}

function makeStreamCollector(filePath) {
  const stream = createWriteStream(filePath, { flags: "w" });
  let text = "";
  let capturedBytes = 0;

  return {
    get text() {
      return text;
    },
    write(chunk) {
      stream.write(chunk);
      if (capturedBytes >= CAPTURE_LIMIT_BYTES) {
        return;
      }
      const remaining = CAPTURE_LIMIT_BYTES - capturedBytes;
      const piece = chunk.length > remaining ? chunk.subarray(0, remaining) : chunk;
      text += piece.toString("utf8");
      capturedBytes += piece.length;
    },
    close() {
      return new Promise((resolve) => {
        stream.end(resolve);
      });
    },
  };
}

function extractFinalAnswer(stdout, stderr) {
  const whole = parseJson(stdout.trim());
  const wholeText = whole === null ? null : findTextCandidate(whole);
  if (wholeText) {
    return { content: wholeText, strategy: "json_stdout" };
  }

  const lines = stdout.split(/\r?\n/).filter((line) => line.trim().length > 0);
  for (let index = lines.length - 1; index >= 0; index -= 1) {
    const value = parseJson(lines[index].trim());
    if (value === null) {
      continue;
    }
    const text = findTextCandidate(value);
    if (text) {
      return { content: text, strategy: "jsonl_stdout" };
    }
  }

  const stdoutFallback = stdout.trim();
  if (stdoutFallback) {
    return { content: stdoutFallback, strategy: "stdout_fallback" };
  }

  const stderrFallback = stderr.trim();
  if (stderrFallback) {
    return { content: stderrFallback, strategy: "stderr_fallback" };
  }

  return { content: "", strategy: "empty" };
}

function findTextCandidate(value) {
  if (value === null || value === undefined) {
    return null;
  }
  if (typeof value === "string") {
    const trimmed = value.trim();
    return trimmed || null;
  }
  if (Array.isArray(value)) {
    const parts = value
      .map((item) => findTextCandidate(item))
      .filter((item) => item && item.trim().length > 0);
    return parts.length > 0 ? parts.join("\n") : null;
  }
  if (typeof value !== "object") {
    return null;
  }

  for (const key of [
    "final_answer",
    "finalAnswer",
    "answer",
    "reply",
    "response",
    "text",
    "content",
    "output",
    "message",
    "result",
    "summary",
  ]) {
    if (Object.hasOwn(value, key)) {
      const text = findTextCandidate(value[key]);
      if (text) {
        return text;
      }
    }
  }

  if (Array.isArray(value.choices)) {
    const text = findTextCandidate(value.choices);
    if (text) {
      return text;
    }
  }

  if (Array.isArray(value.messages)) {
    for (let index = value.messages.length - 1; index >= 0; index -= 1) {
      const message = value.messages[index];
      if (!message || typeof message !== "object") {
        continue;
      }
      if (!message.role || String(message.role).toLowerCase() === "assistant") {
        const text = findTextCandidate(message);
        if (text) {
          return text;
        }
      }
    }
  }

  for (const nestedValue of Object.values(value)) {
    const text = findTextCandidate(nestedValue);
    if (text) {
      return text;
    }
  }

  return null;
}

function parseJson(raw) {
  if (!raw) {
    return null;
  }
  try {
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

function redactCommandForTrace(command, args) {
  const redacted = [command];
  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    redacted.push(arg);
    if (arg === "--message") {
      redacted.push("<task prompt>");
      index += 1;
    }
  }
  return redacted;
}

function normalizeExitCode(exitCode, signal) {
  if (Number.isInteger(exitCode)) {
    return exitCode;
  }
  if (signal) {
    return 128;
  }
  return 1;
}

function ensureTrailingNewline(value) {
  return value.endsWith("\n") ? value : `${value}\n`;
}

async function readJsonIfExists(filePath, fallback) {
  try {
    return JSON.parse(await fs.readFile(filePath, "utf8"));
  } catch (error) {
    if (error.code === "ENOENT") {
      return fallback;
    }
    return fallback;
  }
}

async function fileExists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

async function ensureDir(dirPath) {
  await fs.mkdir(dirPath, { recursive: true });
}

async function appendTrace(tracePath, payload) {
  await fs.appendFile(
    tracePath,
    JSON.stringify(
      {
        timestamp: new Date().toISOString(),
        ...payload,
      },
      null,
      0,
    ) + "\n",
    "utf8",
  );
}

main().catch(async (error) => {
  const outputFlagIndex = process.argv.indexOf("--output");
  if (outputFlagIndex >= 0 && process.argv[outputFlagIndex + 1]) {
    const outputDir = path.resolve(process.argv[outputFlagIndex + 1]);
    await ensureDir(outputDir);
    const message = `${error.name || "Error"}: ${error.message || String(error)}`;
    await fs.writeFile(path.join(outputDir, "final_answer.md"), `${message}\n`, "utf8");
    await appendTrace(path.join(outputDir, "trace.jsonl"), {
      type: "adapter_failed",
      adapter: ADAPTER_NAME,
      error: message,
    });
  }
  process.stderr.write(`${error.stack || error}\n`);
  process.exit(1);
});

File: docker/hermes-runner/adapter/run_task.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
from datetime import datetime, timezone
import json
import os
from pathlib import Path
import re
import shutil
import subprocess
import sys
import time
from typing import Any


ADAPTER_NAME = "nanoclaw-hermes-adapter"


def main(argv: list[str] | None = None) -> int:
    options = parse_args(argv)
    options.output.mkdir(parents=True, exist_ok=True)
    options.state.mkdir(parents=True, exist_ok=True)
    options.workspace.mkdir(parents=True, exist_ok=True)

    trace_path = options.output / "trace.jsonl"
    started_at = time.monotonic()
    task_text = options.task.read_text(encoding="utf-8")
    input_dir = options.task.parent
    runner_request = read_json(input_dir / "runner_request.json", {})
    prior_messages = read_json(input_dir / "prior_messages.json", [])
    resolved_task = read_json(options.resolved_task, {})

    env = build_hermes_env(options, runner_request)
    message = build_message(task_text, prior_messages, env)
    command = env.get("NANOCLAW_HERMES_CLI", "hermes")
    args = build_hermes_args(message, env)
    stdout_path = options.output / "hermes_stdout.log"
    stderr_path = options.output / "hermes_stderr.log"

    append_trace(
        trace_path,
        {
            "type": "adapter_started",
            "adapter": ADAPTER_NAME,
            "task_id": runner_request.get("task_id") or resolved_task.get("id"),
            "run_id": runner_request.get("run_id"),
            "turn": runner_request.get("turn"),
            "mode": hermes_mode(env),
        },
    )
    append_trace(
        trace_path,
        {
            "type": "hermes_started",
            "command": redact_command(command, args),
            "cwd": str(options.workspace),
            "hermes_home": env.get("HERMES_HOME"),
        },
    )

    result = run_command(
        command,
        args,
        cwd=options.workspace,
        env=env,
        stdout_path=stdout_path,
        stderr_path=stderr_path,
    )
    extraction = extract_final_answer(result.stdout, result.stderr)
    final_answer = ensure_trailing_newline(extraction["content"])
    (options.output / "final_answer.md").write_text(final_answer, encoding="utf-8")
    state_cleanup = cleanup_hermes_state(env, options.state)

    metadata = {
        "adapter": ADAPTER_NAME,
        "framework": "hermes-agent",
        "command": redact_command(command, args),
        "cwd": str(options.workspace),
        "hermes_home": env.get("HERMES_HOME"),
        "mode": hermes_mode(env),
        "exit_code": result.returncode,
        "final_answer_strategy": extraction["strategy"],
        "final_answer_bytes": len(final_answer.encode("utf-8")),
        "stdout_file": stdout_path.name,
        "stderr_file": stderr_path.name,
        "state_cleanup": state_cleanup,
        "duration_ms": int((time.monotonic() - started_at) * 1000),
    }
    (options.output / "runner_metadata.json").write_text(
        json.dumps(metadata, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    append_trace(
        trace_path,
        {
            "type": "hermes_finished",
            "exit_code": result.returncode,
            "final_answer_strategy": extraction["strategy"],
            "state_cleanup": state_cleanup,
            "duration_ms": metadata["duration_ms"],
        },
    )
    return result.returncode


def parse_args(argv: list[str] | None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Run a nanoclaw task with Hermes Agent.")
    parser.add_argument("--task", required=True, type=Path)
    parser.add_argument("--resolved-task", required=True, type=Path, dest="resolved_task")
    parser.add_argument("--workspace", required=True, type=Path)
    parser.add_argument("--output", required=True, type=Path)
    parser.add_argument("--state", required=True, type=Path)
    return parser.parse_args(argv)


def build_hermes_env(options: argparse.Namespace, runner_request: dict[str, Any]) -> dict[str, str]:
    env = dict(os.environ)
    env.setdefault("HERMES_HOME", str(options.state))
    env.setdefault("HOME", str(Path(env["HERMES_HOME"]) / "home"))
    env.setdefault("XDG_CONFIG_HOME", str(Path(env["HERMES_HOME"]) / "xdg" / "config"))
    env.setdefault("XDG_STATE_HOME", str(Path(env["HERMES_HOME"]) / "xdg" / "state"))
    env.setdefault("XDG_CACHE_HOME", str(Path(env["HERMES_HOME"]) / "xdg" / "cache"))

    if not env.get("NANOCLAW_MODEL") and runner_request.get("model"):
        env["NANOCLAW_MODEL"] = str(runner_request["model"])
    if env.get("NANOCLAW_MODEL") and not env.get("HERMES_INFERENCE_MODEL"):
        env["HERMES_INFERENCE_MODEL"] = env["NANOCLAW_MODEL"]
    if env.get("NANOCLAW_HERMES_MODEL") and not env.get("HERMES_INFERENCE_MODEL"):
        env["HERMES_INFERENCE_MODEL"] = env["NANOCLAW_HERMES_MODEL"]
    if env.get("NANOCLAW_HERMES_PROVIDER") and not env.get("HERMES_INFERENCE_PROVIDER"):
        env["HERMES_INFERENCE_PROVIDER"] = env["NANOCLAW_HERMES_PROVIDER"]
    if env.get("NANOCLAW_BASE_URL") and not env.get("OPENAI_BASE_URL"):
        env["OPENAI_BASE_URL"] = env["NANOCLAW_BASE_URL"]
    if env.get("NANOCLAW_BASE_URL") and not env.get("CUSTOM_BASE_URL"):
        env["CUSTOM_BASE_URL"] = env["NANOCLAW_BASE_URL"]
    elif env.get("OPENAI_BASE_URL") and not env.get("CUSTOM_BASE_URL"):
        env["CUSTOM_BASE_URL"] = env["OPENAI_BASE_URL"]
    if env.get("CUSTOM_BASE_URL") and not env.get("HERMES_INFERENCE_PROVIDER"):
        env["HERMES_INFERENCE_PROVIDER"] = "custom"

    for directory in [
        Path(env["HERMES_HOME"]),
        Path(env["HOME"]),
        Path(env["XDG_CONFIG_HOME"]),
        Path(env["XDG_STATE_HOME"]),
        Path(env["XDG_CACHE_HOME"]),
    ]:
        directory.mkdir(parents=True, exist_ok=True)

    return env


def build_message(task_text: str, prior_messages: Any, env: dict[str, str]) -> str:
    if env.get("NANOCLAW_HERMES_INCLUDE_PRIOR_MESSAGES", "1") == "0":
        return task_text
    if not isinstance(prior_messages, list) or not prior_messages:
        return task_text

    history_parts: list[str] = []
    for message in prior_messages:
        if not isinstance(message, dict):
            continue
        role = str(message.get("role") or "message").upper()
        content = str(message.get("content") or "")
        if content.strip():
            history_parts.append(f"[{role}]\n{content}")

    if not history_parts:
        return task_text
    return "\n\n".join(history_parts + [f"[USER]\n{task_text}"])


def hermes_mode(env: dict[str, str]) -> str:
    explicit = env.get("NANOCLAW_HERMES_MODE", "").strip().lower()
    if explicit in {"chat", "oneshot"}:
        return explicit
    chat_only_vars = [
        "NANOCLAW_HERMES_TOOLSETS",
        "NANOCLAW_HERMES_SKILLS",
        "NANOCLAW_HERMES_MAX_TURNS",
        "NANOCLAW_HERMES_SOURCE",
        "NANOCLAW_HERMES_RESUME",
        "NANOCLAW_HERMES_CONTINUE",
    ]
    return "chat" if any(env.get(name) for name in chat_only_vars) else "oneshot"


def build_hermes_args(message: str, env: dict[str, str]) -> list[str]:
    mode = hermes_mode(env)
    args: list[str] = []

    if env.get("NANOCLAW_HERMES_RESUME"):
        args.extend(["--resume", env["NANOCLAW_HERMES_RESUME"]])
    if env.get("NANOCLAW_HERMES_CONTINUE"):
        args.extend(["--continue", env["NANOCLAW_HERMES_CONTINUE"]])
    if env.get("NANOCLAW_HERMES_YOLO") == "1":
        args.append("--yolo")
    if env.get("NANOCLAW_HERMES_IGNORE_USER_CONFIG") == "1":
        args.append("--ignore-user-config")
    if env.get("NANOCLAW_HERMES_IGNORE_RULES") == "1":
        args.append("--ignore-rules")

    provider = env.get("NANOCLAW_HERMES_PROVIDER") or env.get("HERMES_INFERENCE_PROVIDER")
    model = (
        env.get("NANOCLAW_HERMES_MODEL")
        or env.get("NANOCLAW_MODEL")
        or env.get("HERMES_INFERENCE_MODEL")
    )

    if mode == "chat":
        args.extend(["chat", "--quiet", "-q", message])
        if env.get("NANOCLAW_HERMES_SOURCE"):
            args.extend(["--source", env["NANOCLAW_HERMES_SOURCE"]])
        if env.get("NANOCLAW_HERMES_TOOLSETS"):
            args.extend(["--toolsets", env["NANOCLAW_HERMES_TOOLSETS"]])
        if env.get("NANOCLAW_HERMES_SKILLS"):
            args.extend(["--skills", env["NANOCLAW_HERMES_SKILLS"]])
        if env.get("NANOCLAW_HERMES_MAX_TURNS"):
            args.extend(["--max-turns", env["NANOCLAW_HERMES_MAX_TURNS"]])
    else:
        args.extend(["-z", message])

    if provider:
        args.extend(["--provider", provider])
    if model:
        args.extend(["--model", model])
    return args


def run_command(
    command: str,
    args: list[str],
    *,
    cwd: Path,
    env: dict[str, str],
    stdout_path: Path,
    stderr_path: Path,
) -> subprocess.CompletedProcess[str]:
    try:
        result = subprocess.run(
            [command, *args],
            check=False,
            cwd=str(cwd),
            env=env,
            capture_output=True,
            text=True,
            errors="replace",
        )
    except FileNotFoundError as exc:
        result = subprocess.CompletedProcess(
            [command, *args],
            returncode=127,
            stdout="",
            stderr=str(exc),
        )

    stdout_path.write_text(result.stdout or "", encoding="utf-8")
    stderr_path.write_text(result.stderr or "", encoding="utf-8")
    return result


def cleanup_hermes_state(env: dict[str, str], state_dir: Path) -> dict[str, Any]:
    skills_dir = Path(env.get("HERMES_HOME") or state_dir) / "skills"
    payload: dict[str, Any] = {
        "skills_dir": str(skills_dir),
        "removed_skills": False,
    }
    if env.get("NANOCLAW_HERMES_KEEP_STATE_SKILLS") == "1":
        payload["skipped_reason"] = "disabled_by_env"
        return payload

    if not skills_dir.exists():
        payload["skipped_reason"] = "missing"
        return payload

    try:
        resolved_skills_dir = skills_dir.resolve(strict=False)
        resolved_state_dir = state_dir.resolve(strict=False)
        if not resolved_skills_dir.is_relative_to(resolved_state_dir):
            payload["skipped_reason"] = "outside_state_dir"
            return payload
    except OSError as exc:
        payload["skipped_reason"] = "path_resolution_failed"
        payload["error"] = str(exc)
        return payload

    try:
        shutil.rmtree(skills_dir)
    except OSError as exc:
        payload["skipped_reason"] = "remove_failed"
        payload["error"] = str(exc)
        return payload

    payload["removed_skills"] = True
    return payload


def extract_final_answer(stdout: str, stderr: str) -> dict[str, str]:
    stdout_clean = strip_ansi(stdout).strip()
    stderr_clean = strip_ansi(stderr).strip()
    if stdout_clean:
        return {"content": stdout_clean, "strategy": "stdout"}
    if stderr_clean:
        return {"content": stderr_clean, "strategy": "stderr_fallback"}
    return {"content": "", "strategy": "empty"}


def strip_ansi(value: str) -> str:
    return re.sub(r"\x1b\[[0-?]*[ -/]*[@-~]", "", value)


def redact_command(command: str, args: list[str]) -> list[str]:
    redacted = [command]
    redact_next = False
    for arg in args:
        if redact_next:
            redacted.append("<task prompt>")
            redact_next = False
            continue
        redacted.append(arg)
        if arg in {"-z", "-q", "--query"}:
            redact_next = True
    return redacted


def read_json(path: Path, fallback: Any) -> Any:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (FileNotFoundError, json.JSONDecodeError):
        return fallback


def append_trace(path: Path, payload: dict[str, Any]) -> None:
    event = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        **payload,
    }
    with path.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(event, ensure_ascii=False) + "\n")


def ensure_trailing_newline(value: str) -> str:
    return value if value.endswith("\n") else f"{value}\n"


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        output_dir = None
        if "--output" in sys.argv:
            index = sys.argv.index("--output")
            if index + 1 < len(sys.argv):
                output_dir = Path(sys.argv[index + 1])
        if output_dir is not None:
            output_dir.mkdir(parents=True, exist_ok=True)
            message = f"{type(exc).__name__}: {exc}"
            (output_dir / "final_answer.md").write_text(message + "\n", encoding="utf-8")
            append_trace(
                output_dir / "trace.jsonl",
                {
                    "type": "adapter_failed",
                    "adapter": ADAPTER_NAME,
                    "error": message,
                },
            )
        raise
